src/field/HelixStepper.hh

```diff
@@ -0,0 +1,164 @@
+// See the top-level COPYRIGHT file for details.
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file HelixStepper.hh
+//---------------------------------------------------------------------------//
+#pragma once
+
+#include "base/Types.hh"
+
+#include "Types.hh"
+
+namespace celeritas
+{
+//---------------------------------------------------------------------------//
+/*!
+ * The Helix Stepper
+ *
+ * The analytical solution for the motion of a charged particle in a uniform
+ * magnetic field.
+ */
+template<class EquationT>
+class HelixStepper
+{
+  public:
+    //!@{
+    //! Type aliases
+    using Result = StepperResult;
+    //!@}
+
+  public:
+    // Construct with the equation of motion
+    CELER_FUNCTION
+    HelixStepper(const EquationT& eq) : equation_(eq) {}
+
+    // Adaptive step size control
+    CELER_FUNCTION auto operator()(real_type step, const OdeState& beg_state)
+        -> Result;
+
+  private:
+    //// DATA ////
+
+    // Equation of the motion
+    const EquationT& equation_;
+
+    //// HELPER FUNCTIONS ////
+
+    // Analytical solution for a given step along a helix trajectory
+    CELER_FUNCTION OdeState move(real_type       step,
+                                 real_type       radius,
+                                 real_type       helicity,
+                                 const OdeState& beg_state,
+                                 const OdeState& rhs);
+
+    //// COMMON PROPERTIES ////
+
+    static CELER_CONSTEXPR_FUNCTION real_type tolerance() { return 1e-10; }
+};
+
+//---------------------------------------------------------------------------//
+// INLINE DEFINITIONS
+//---------------------------------------------------------------------------//
+/*!
+ * An explicit helix stepper with analytical solutions at the end and the
+ * middle point for a given step.
+ *
+ * Assuming that the magnetic field is uniform and chosen to be parallel to
+ * the z-axis, \f$B = (0, 0, B_z)\f$, without loss of generality, the motion of
+ * a charged particle is described by a helix trajectory. For this algorithm,
+ * the radius of the helix, \f$R = m gamma v/(qB)\f$ and the helicity, defined
+ * as \f$ -sign(q B_z)\f$ are evaluated through the right hand side of the ODE
+ * equation where q is the charge of the particle.
+ */
+template<class E>
+CELER_FUNCTION auto
+HelixStepper<E>::operator()(real_type step, const OdeState& beg_state)
+    -> Result
+{
+    Result result;
+
+    // Evaluate the right hand side of the equation
+    OdeState rhs = equation_(beg_state);
+
+    // Calculate the radius of the helix
+    real_type radius
+        = sqrt(ipow<2>(norm(beg_state.mom)) - ipow<2>(beg_state.mom[2]))
+          / norm(rhs.mom);
+
+    // Set the helicity: 1(-1) for negative(positive) charge with Bz > 0
+    real_type helicity = (rhs.mom[0] / rhs.pos[1] < 0) ? 1 : -1;
```

Comment:
I think a `bool` (or better, `enum class Helicity : bool { positive, negative };` rather than a 64-bit real would be more suitable here. That plus a helper function `int to_sign(Helicity h) { return static_cast<int>(h) * 2 - 1; }` would give you branch-free arithmetic with lower register usage.